api/src/main/java/uk/gov/hmcts/payment/api/controllers/OrderController.java

```diff
@@ -0,0 +1,254 @@
+            } catch (JsonProcessingException e) {
+                return new ResponseEntity<>(e.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
+            }
+            return new ResponseEntity<>(responseBO, HttpStatus.valueOf(idempotencyKeys.getResponseCode())); // if hashcode matched
+        };
+
+        //Idempotency Check
+        Optional<IdempotencyKeys> idempotencyKeysRow = getIdempotencyKey.apply(idempotencyKey);
+        if (idempotencyKeysRow.isPresent()) {
+            ResponseEntity responseEntity = validateHashcodeForRequest.apply(idempotencyKeysRow.get());
+            return responseEntity;
+        }
+
+        //business validations for order
+        PaymentFeeLink order = businessValidationForOrders(orderDomainService.find(orderReference), orderPaymentDto);
+
+        //PBA Payment
+        OrderPaymentBo orderPaymentBo = null;
+        ResponseEntity responseEntity;
+        String responseJson;
+        try {
+            orderPaymentBo = orderDomainService.addPayments(order, orderPaymentDto);
+            HttpStatus httpStatus = orderPaymentBo.getStatus().equalsIgnoreCase(FAILED) ? HttpStatus.PAYMENT_REQUIRED : HttpStatus.CREATED; //402 for failed Payment scenarios
+            responseEntity = new ResponseEntity<>(orderPaymentBo, httpStatus);
+            responseJson = objectMapper.writeValueAsString(orderPaymentBo);
+        } catch (LiberataServiceTimeoutException liberataServiceTimeoutException) {
+            responseEntity = new ResponseEntity<>(liberataServiceTimeoutException.getMessage(), HttpStatus.GATEWAY_TIMEOUT);
+            responseJson = liberataServiceTimeoutException.getMessage();
+        }
+
+        //Create Idempotency Record
+        return createIdempotencyRecord(objectMapper, idempotencyKey, orderReference, responseJson, responseEntity, orderPaymentDto);
+    }
+
+    private PaymentFeeLink businessValidationForOrders(PaymentFeeLink order, OrderPaymentDto orderPaymentDto) {
+        //Business validation for amount
+        Optional<BigDecimal> totalCalculatedAmount = order.getFees().stream().map(paymentFee -> paymentFee.getCalculatedAmount()).reduce(BigDecimal::add);
+        if (totalCalculatedAmount.isPresent() && (totalCalculatedAmount.get().compareTo(orderPaymentDto.getAmount()) != 0)) {
+            throw new OrderExceptionForNoMatchingAmount("The order amount should be equal to order balance");
+        }
+
+
+        //Business validation for amount due for fees
+        Optional<BigDecimal> totalAmountDue = order.getFees().stream().map(paymentFee -> paymentFee.getAmountDue()).reduce(BigDecimal::add);
+        if (totalAmountDue.isPresent() && totalAmountDue.get().compareTo(BigDecimal.ZERO) == 0) {
+            throw new OrderExceptionForNoAmountDue("The order has already been paid");
+        }
+
+        return order;
+    }
+
+
+    private ResponseEntity createIdempotencyRecord(ObjectMapper objectMapper, String idempotencyKey, String orderReference,
```

Comment:
can't we move this method into service layer? Ideally, Controller classes should only have endpoints in them

Reply:
Moved